### PR TITLE
Send Covid-19 delay email to candidates

### PR DESCRIPTION
**Pull request by malcolmbaig**
## Context

Need to update candidates with new guidance relating to COVID 19 so candidates are aware of the impacts to timeframes 

## Changes proposed in this pull request

Screenshot 2020-03-23 at 17 19 21

## Guidance to review

* Email preview available locally.

## Link to Trello card

https://trello.com/c/yDzH0Tfv

## Things to check

- [x] This code doesn't rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training#azure-hosting-devops-pipeline)

### CODE REVIEW

File: app/lib/covid19/candidate_email_delivery.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+      candidates = Candidate.includes(application_forms: [:application_choices])
+
+      candidates_to_email = candidates.select do |candidate|
+        application_form = candidate.current_application
+        if application_form.application_choices.all? { |ac| ac.status.in? skip_email_statuses }
+          false
+        else
+          true
+        end
+      end
```
Comment (tvararu) [inline app/lib/covid19/candidate_email_delivery.rb]: This is a Ruby array `.select` right? So it'll be slower than running in the database. We've only got 500 candidates on prod so I don't think this is likely to be a bottleneck...

File: app/views/covid19_mailer/covid_19_delay.text.erb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+Dear <%= @application_form.first_name || 'candidate' %>
+
+Given the impact of coronavirus (COVID-19) on schools and universities, your teacher training application might be processed more slowly than usual.
+
+To check the status of your application, sign in to Apply for teacher training https://www.apply-for-teacher-training.education.gov.uk/candidate/sign-in
+
+If you have any questions about this, you can contact us at [becomingateacher@digital.education.gov.uk](mailto:becomingateacher@digital.education.gov.uk)
```
Comment (malcolmbaig) [inline app/views/covid19_mailer/covid_19_delay.text.erb]: Had to type this out as I couldn't see anything other than a screenshot of the copy on the Trello card. Would appreciate a proof-read.

File: app/views/covid19_mailer/covid_19_delay.text.erb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+Dear <%= @application_form.first_name || 'candidate' %>
+
+Given the impact of coronavirus (COVID-19) on schools and universities, your teacher training application might be processed more slowly than usual.
+
+To check the status of your application, sign in to Apply for teacher training https://www.apply-for-teacher-training.education.gov.uk/candidate/sign-in
```
Comment (chubberlisk) [inline app/views/covid19_mailer/covid_19_delay.text.erb]: We have `ViewHelper#candidate_sign_in_url` for generating the sign in link for a candidate, see `app/helpers/view_helper.rb:54`.

```suggestion
To check the status of your application, sign in to <%= candidate_sign_in_url(candidate) %>
```

File: app/mailers/covid19_mailer.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+class Covid19Mailer < ApplicationMailer
+  def covid_19_delay(application_form)
```
Comment (chubberlisk) [inline app/mailers/covid19_mailer.rb]: Nitpick, thinking `covid_19` is already part of the mailer name. Feel free to ignore!

```suggestion
  def delay(application_form)
```

File: app/views/covid19_mailer/covid_19_delay.text.erb (hunk 2 above)
Comment (C-gyorfi) [reply app/views/covid19_mailer/covid_19_delay.text.erb]: Looks good!

Comment (chubberlisk) [reply app/views/covid19_mailer/covid_19_delay.text.erb]: Did a proofread against the screenshot. Looks right to my eyes! 🕵️‍♀️

File: app/lib/covid19/candidate_email_delivery.rb (hunk 1 above)
Comment (malcolmbaig) [reply app/lib/covid19/candidate_email_delivery.rb]: Yep, opted for Ruby as it's arguably clearer than the equivalent SQL and total record numbers are low.

File: app/views/covid19_mailer/covid_19_delay.text.erb (hunk 3 above)
Comment (malcolmbaig) [reply app/views/covid19_mailer/covid_19_delay.text.erb]: Might have better not to introduce an expiring token into the email 

> How are we going to trigger this mailer? Command line?

Yep, was thinking just use the rails runner on an Azure shell.

**Comment by malcolmbaig** [comment]
> How are we going to trigger this mailer? Command line?

Yep, can just use `rails runner` on an Azure shell.

File: app/views/covid19_mailer/covid_19_delay.text.erb (hunk 3 above)
Comment (malcolmbaig) [reply app/views/covid19_mailer/covid_19_delay.text.erb]: 👍 , updated.

File: app/mailers/covid19_mailer.rb (hunk 4 above)
Comment (malcolmbaig) [reply app/mailers/covid19_mailer.rb]: Agreed, reworked the method name and the location of the preview code.

**Pull request merged**